### ISSUE TITLE
Feature: prompting to save a launch configuration for scheduled & workflow nodes

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#169]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
 - API
 - UI

##### SUMMARY

Currently, we allow prompting for many things at job launch time. Any of the following could be blank and need to be supplied:

- extra_vars (in general)
- extra_vars (via surveys)
- tags/skip_tags
- inventory
- credential

etc.

We should be able to have this launch configuration saved for schedule definitions and workflow nodes. Can re-use the launch dialog to populate this.

Bonus points for a 'schedule this job' button on job details as well.

##### ADDITIONAL INFORMATION

This is *not* about bringing prompting to the top level of the workflow itself. That is a different item (because it brings complication about precedence, overriding, etc.)

**Comment by MrMEEE** [comment]
+1

**Comment by AlanCoding** [comment]
I think the ideal way to handle this would be to repurpose the models `WorkflowJobTemplateNode` and `WorkflowJobNode` to be a generalized "launch time configuration", which means rearchitecting how prompting works.

If anyone is confused about why I say this, have a look at the `JobAccess` relaunch method. We have already dealt with implicit requests to separate out what parameters were given by the user versus what came from the template _after_ a job has been launched.

I don't know if we'll be able to pull it off quite like that (renaming a model might not be viable), but I would imagine some kind of top-level modification or addition to the models in the system.

**Comment by wwitzel3** [comment]
Notes from planning session

Places that indirectly launch jobs, such as Schedules and Workflows, should be able to provide the answers to prompt-able fields. Example: reduce needing to create a new JT for each inventory/credential that you want to run a scheduled job for.

* Add prompt-able fields at node creation/edit in workflow editor.
  * Anything that is prompt on launch should be able to be filled in.
* Relax defaults being required for prompt fields in workflow node.
  * Validate the workflow prompt fields on save in workflow editor if they don't have defaults.

Credentials
=======
Show prompt for password items when creating scheduled jobs / workflows, but mark them as incompatible with scheduled jobs / workflow jobs. API will provide this detail so the UI can present it.

Prompt on launch for a credential is allowed, but the credential can not be prompt for password.

Implementation Thoughts
============
Abstract node class / mixin that Schedule inherits. Schedule becomes a node.

**Comment by AlanCoding** [comment]
Migrating the final bits of what I have down:

----
### Out of Scope
~TODO:~ DONE: Make new issue for workflow ask extra vars on launch
(if circumstances demand) Could look into saving cred password prompts for scheduled jobs

----
### In Scope


#### Side Items

Expose new field on credentials like `can_use_without_user_input` for the UI to show it can/can’t be used in WFJT. It can't be used in a WFJT if it requires user password prompt.

#### Major Development Areas

WFJT nodes:

 - Extra_vars
    - API: add WFJT node extra_vars field w/encrypted values
    - UI: If JT has survey, provide menu to answer survey questions
    - If JT is set to ask variables, can provide variables free-form
    - survey password-type answers should be encrypted and never shown

Schedules:

 - What is promptable?
    - project/inv updates, sys. jobs: NOTHING
    - WFJTs: vars
    - JTs: lots
 - Proposed design - add fields on Schedule model for anything JT launch accepts
    - extra_data (remains)
       - UI: If JT has survey, provide menu to answer survey questions (here too)
    - Inventory
    - Credential
    - ...and all the rest...

Job “schedule this job”:
 - New API endpoint /schedule_job/, will create new schedule entry
    - New schedule is disabled by default
    - New schedule Has no recurrence rule (until user sets it)
    - New schedule has all same "prompted" fields as what user provided on launch
    - Validation
       - Will not work if credential passwords prompted for
 - UI will have button to click, POST to endpoint, navigate to schedule edit view

**Comment by AlanCoding** [comment]
> TODO: Make new issue for workflow ask extra vars on launch

https://github.com/ansible/awx/issues/478

**Comment by mabashian** [comment]
Workflow UI Implication Notes:

- Move Always/Successful/Failure radio buttons to RUN dropdown (see mockups)
- Add tooltip to SELECT button when user needs to fill in promptable fields
- Add prompt button which leverages existing launch modal for filling out promptable fields/survey
- Add PREVIEW tab to launch modal
- Need to be able to handle the case where a user starts filling out prompts and doesn't finish
- Should we show a read-only aggregated extra vars codemirror type thing after a node has been created?
- Potentially explore workflow layout tweaks to gain a bit more space
- Stop filtering out credentials that prompt for password on launch (we won’t be supporting these but we shouldn’t filter them out.  Should let the user know that this credential is not compatible with workflow nodes).

Schedule UI Implication Notes:

- Add tooltip to SAVE button when user needs to fill in promptable fields
- Add prompt button which leverages existing launch modal for filling out promptable fields/survey
- Need to be able to handle the case where a user starts filling out prompts and doesn't finish
- Bonus: Schedule JT run from job details screen which takes the user to the edit schedule form -> disabled by default, no recurrence.  Does this also apply to workflows?

**Comment by AlanCoding** [comment]
@wenottingham we've discovered some new questions about credential behavior, two components:

 - say user provides `"credentials": []` on launch, should this remove existing credentials from the JT?
 - say a user provides `"credentials": [1, 2]` where the JT has attached `[3, 4]`. What should the job launch with?

The precedent we have set (given the history of `extra_credentials`) is that, in the first case, we launch without the JT credentials, and in the 2nd case, we launch with `[1, 2]`.

With this release, launching with `"credentials": []` leaves us without a machine credential, and thus is an incoherent request. This might be an issue with the current multi-credential work, it should probably be tested.

But in the future, we're going to allow running job templates without machine credentials. There might be an eventual request to use `ask_credential_on_launch` and have a saved launch-time configuration launch the job without any credentials, in spite of the fact that the JT has credentials. This is a problem, because if we use a many-to-many relationship (being congruent with the JT API), there is no way to distinguish between "blow away all credentials on launch" and "do not change credentials on launch".

What use-cases are going to be important in future features? If we plan to allow save launch-time configurations to null-out credentials, this will involve substantial rearchitecting of the approach being taken now, suffering other tradeoffs. If that's not an important scenario, then we can continue on the current path pretty well unchanged.
